pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -232,27 +232,33 @@ func (r *LazyBinaryReader) load() (returnErr error) {
 		return r.readerErr
 	}
 
+	// Release the read lock, so that load1 can take write lock. Take the read lock again once done.
+	r.readerMx.RUnlock()
+	err := r.load1()
+	r.readerMx.RLock()
+
+	// Between the write lock release and the subsequent read lock, the unload() may have run,
+	// so we make sure to catch this edge case.
+	if err == nil && r.reader == nil {
+		err = errUnloadedWhileLoading
+	}
+	return err
+}
+
+// load1 is called from load, after releasing the read lock. load1 will acquire write lock instead.
+func (r *LazyBinaryReader) load1() error {
```

Comment:
[nit] Nice moving this logic to a dedicated function. Have you considered an alternative naming, like `loadWithGate()`?

Reply:
Please see follow-up PR: https://github.com/grafana/mimir/pull/5608